### PR TITLE
Array destructuring

### PR DESCRIPTION
**Pull request by romdotdog**
- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

This is a ~~work in progress~~ ~~almost finished~~ finished PR to hopefully implement array destructuring

### CODE REVIEW

File: src/compiler.ts [hunk 1]
```diff
@@ -2999,195 +3000,325 @@ export class Compiler extends DiagnosticEmitter {
...
+                      name
+                    );
+                    continue;
+                  }
+                  local = flow.parentFunction.addLocal(type, name, declaration);
+                  if (isConst) flow.setLocalFlag(local.index, LocalFlags.CONSTANT | LocalFlags.INITIALIZED);
+                }
+
+                let callExpr = this.compileCallDirect(indexedGet, [
+                  Node.createIntegerLiteralExpression(i64_new(j), nameNode.range)
```
Comment (romdotdog) [inline src/compiler.ts:3116]: Is there a better way to do this?

**Comment by romdotdog** [comment]
Should I add rest parameters? If so, should I rely on there being a `slice` method like in 
https://github.com/AssemblyScript/assemblyscript/issues/1473#issuecomment-699172441

**Comment by trusktr** [comment]
:partying_face:

**Comment by romdotdog** [comment]
Parser test `parser/continue-on-error.ts` was failing, now fixed

**Comment by romdotdog** [comment]
To reduce the size of this PR, I will not be including rest parameters or array destructuring in function parameters
This PR is now generally ready for review and merge, I will meanwhile make sure that tests pass

File: src/extra/ast.ts [hunk 2]
```diff
@@ -613,13 +638,13 @@ export class ASTBuilder {
   visitFunctionExpression(node: FunctionExpression): void {
     var declaration = node.declaration;
     if (!declaration.arrowKind) {
-      if (declaration.name.text.length) {
+      if (declaration.ident.text.length) {
```
Comment (MaxGraey) [inline src/extra/ast.ts:641]: Hmm, I'm not sure `ident` is good name. It could be misreading as `indent`. Why not keep the old `name`?

**Comment by MaxGraey** [comment]
Does swapping support?
```ts
let a = 1;
let b = 3;
[a, b] = [b, a];
assert(a == 3);
assert(b == 1);
```

File: src/extra/ast.ts (hunk 2 above)
Comment (romdotdog) [reply src/extra/ast.ts:641]: What should the old `name` (that is still in use) become?

**Comment by romdotdog** [comment]
Yes, it does.

File: src/ast.ts [hunk 3]
```diff
@@ -2045,7 +2073,7 @@ export class FunctionDeclaration extends DeclarationStatement {
   /** Clones this function declaration. */
   clone(): FunctionDeclaration {
     return new FunctionDeclaration(
-      this.name,
+      this.name as IdentifierExpression,
```
Comment (dcodeIO) [inline src/ast.ts:2076]: This can fail silently when compiled to JS. Adding an `assert` above to make sure that the `kind` matches would unify the behavior.

Can you explain why `name` needs to be `Expression`?

Comment (romdotdog) [reply src/ast.ts:2076]: It needs to be able to hold BindingPatternExpression

Comment (romdotdog) [reply src/ast.ts]: It's really a `BindingPatternExpression | IdentifierExpression`

File: src/extra/ast.ts (hunk 2 above)
Comment (MaxGraey) [reply src/extra/ast.ts:641]: I mean revert it to `declaration.name`

File: src/ast.ts (hunk 3 above)
Comment (dcodeIO) [reply src/ast.ts]: Hmm, seems odd that this propagates up to `FunctionDeclaration`. Is there maybe a way to separate this?

Comment (romdotdog) [reply src/ast.ts]: no, because `DeclarationStatement` always holds the property `name`
I recall that @jtenner also thinks this is the only way in a voice chat we had

File: src/extra/ast.ts (hunk 2 above)
Comment (romdotdog) [reply src/extra/ast.ts:641]: > 
> 
> I mean revert it to `declaration.name`

Can't, `ident` is a getter function, `name` is still a property of DeclarationStatement

File: src/ast.ts (hunk 3 above)
Comment (dcodeIO) [reply src/ast.ts]: What if the `name` property would become a getter that can be overridden with a more specific type where necessary, say with a private `_name` that is `Expression`? Would that work?

Comment (romdotdog) [reply src/ast.ts]: That's a good idea, I'll implement that after I have breakfast

File: src/extra/ast.ts (hunk 2 above)
Comment (romdotdog) [reply src/extra/ast.ts:641]: Solved by discussion below

Comment (MaxGraey) [reply src/extra/ast.ts:641]: I see. How about `id` of full name `identifier` instead `ident`?

File: src/ast.ts [hunk 4]
```diff
@@ -1701,6 +1724,17 @@ export abstract class DeclarationStatement extends Statement {
   isAny(flag: CommonFlags): bool { return (this.flags & flag) != 0; }
   /** Sets a specific flag or flags. */
   set(flag: CommonFlags): void { this.flags |= flag; }
+
+  /** Attempts to return the name as an identifier. */
+  get ident(): IdentifierExpression {
+    assert(this._name.kind == NodeKind.IDENTIFIER || this._name.kind == NodeKind.CONSTRUCTOR);
+    return this._name as IdentifierExpression;
```
Comment (romdotdog) [inline src/ast.ts:1731]: any better name for this function?

File: src/program.ts [hunk 5]
```diff
@@ -2897,7 +2927,7 @@ export abstract class DeclaredElement extends Element {
 
   /** Gets the associated identifier node. */
   get identifierNode(): IdentifierExpression {
-    return this.declaration.name;
+    return this.declaration.ident;
   }
```
Comment (romdotdog) [inline src/program.ts:2931]: This seems wrong

File: src/program.ts [hunk 6]
```diff
@@ -2571,6 +2569,27 @@ export class Program extends DiagnosticEmitter {
     parent.add(name, element); // reports
   }
 
+  /** Initializes a variable that's part of a declaration. */
+  private initializeVariable(
+    /** Name of the variable to initialize. */
+    name: string,
+    /** Parent element, usually a file or namespace. */
```
Comment (romdotdog) [inline src/program.ts:2576]: I should probably inline this function again

**Comment by MaxGraey** [comment]
Add please tests for empty destructors:
```ts
const arr = [1, 2, 3];
const [] = arr;
```
and
```ts
const arr = [1, 2, 3];
const [,] = arr;
```

**Comment by romdotdog** [comment]
good catch, thank you

**Comment by MaxGraey** [comment]
Fo next PR will be great also add default assignment as well:
```ts
const [a = 1] = [];
assert(a == 1);
```

**Comment by romdotdog** [comment]
@MaxGraey 

> For now without default value this `const [a] = [] as i32[];` should throw OOB trap
It does. Is there a way integrate this into the test? (test for traps?)

Also, in general, does this PR look good to you?

**Comment by MaxGraey** [comment]
It look good to me but better if @dcodeIO review this

**Comment by MaxGraey** [comment]
@romdotdog btw in this PR `package-lock.json` was changed. Could you remove this unnecessary changes?

**Comment by romdotdog** [comment]
That's weird, it seems the features test failed because of the nvm GitHub action.
cc: @dcodeIO

**Comment by MaxGraey** [comment]
> That's weird, it seems the features test failed because of the nvm GitHub action.

Yeah. That's sometimes happening with nightly builds. Don't care about this. Just need some time

**Comment by MaxGraey** [comment]
> That's weird, it seems the features test failed because of the nvm GitHub action.
> cc: @dcodeIO

please sync with main

File: src/compiler.ts (hunk 1 above)
Comment (dcodeIO) [reply src/compiler.ts:3135]: Perhaps `makeCallDirect` to circumvent the AST and provide a Binaryen IR node directly :)

File: src/ast.ts (hunk 4 above)
Comment (dcodeIO) [reply src/ast.ts:1731]: Perhaps `identifier`, we don't abbreviate so much elsewhere if I'm not mistaken.

File: src/ast.ts [hunk 7]
```diff
@@ -55,6 +55,7 @@ export enum NodeKind {
 
   // expressions
   IDENTIFIER,
+  PATTERN,
```
Comment (dcodeIO) [inline src/ast.ts:58]: What about `BINDINGPATTERN`? :)

Comment (romdotdog) [reply src/ast.ts]: Done @dcodeIO

**Comment by romdotdog** [comment]
> 
> 
> > That's weird, it seems the features test failed because of the nvm GitHub action.
> > cc: @dcodeIO
> 
> please sync with main

Are you talking to me? Is there any trouble with merging this? GitHub is telling me that there are no conflicts :)

**Comment by MaxGraey** [comment]
Yeah, no need, that's why I hid that message. By the way, have you seen the other dcodeIO suggestions?
https://github.com/AssemblyScript/assemblyscript/pull/2008#discussion_r685821303
https://github.com/AssemblyScript/assemblyscript/pull/2008#discussion_r685818957

**Comment by romdotdog** [comment]
Thank you! I completely missed those. I'll implement those tomorrow

**Comment by MaxGraey** [comment]
Could you also add destruct tests for typed and static arrays?
```ts
let arr = new Float32Array(3);
arr[0] = 1;
arr[1] = 2;
arr[2] = 3
let [a, b, c] = arr;
assert(a == 1);
assert(b == 2);
assert(c == 3);
```

**Review by MaxGraey — APPROVED** [review]

File: tests/parser/array-destructuring.ts [hunk 8]
```diff
@@ -0,0 +1,6 @@
+function test1(): void {
+  const [, a, b] = c;
+  const {d, e} = f;
+  const [] = g;
+  const [,] = h;
```
Comment (MaxGraey) [inline tests/parser/array-destructuring.ts:5]: Add please more tests here. Like:
```ts
const [,,  ,
,] = k;
let [
] = n;
var [,a,] = j
```

**Comment by romdotdog** [comment]
BTW, for the next PR I think I'm going to replace binding patterns with array literals for more flexible parsing and for normal assignments and declarations to have more similar code

**Comment by MaxGraey** [comment]
@romdotdog It seems some tests have failed

File: tests/parser/array-destructuring.ts [hunk 9]
```diff
@@ -3,4 +3,21 @@ function test1(): void {
...
+  var [,_,] = j ;
+  var [,,  ,a,, f ,
+    , g,  g,        ] = 1;
+  var [n, a, a, a,v,
+    ,h,h,h,h,h,,r
+    ,,r,,h
+    ,y, ] = 2;
+  let {aaa , a} = 1;
+  var {f,,,,,,,,,,,,} = 3;
+  const {fff} = 3;
```
Comment (MaxGraey) [inline tests/parser/array-destructuring.ts:22]: Btw all that should an error like:
```
TS2488: Type 'number' must have a iterator.
```

**Comment by dcodeIO** [comment]
I like the idea of reusing array literal expressions here, would be cool if it could be done right away :)

File: tests/parser/array-destructuring.ts (hunk 9 above)
Comment (romdotdog) [reply tests/parser/array-destructuring.ts:22]: Yes, but it's a parser test, not a compiler test

**Comment by romdotdog** [comment]
> 
> 
> I like the idea of reusing array literal expressions here, would be cool if it could be done right away :)

It's an easily fixable issue; I don't think this PR is worth delaying more for something that can be included in the next PR (:

File: tests/parser/array-destructuring.ts (hunk 9 above)
Comment (romdotdog) [reply tests/parser/array-destructuring.ts:22]: If this is a request I'll add a test for that but it's unclear what you're asking :)

**Comment by romdotdog** [comment]
BTW, I couldn't reproduce the failed compiler test on my machine.
Not to mention that it failed without me making any changes to the compiler test or the compiler itself.. Maybe someone could point me in the right direction?

**Comment by romdotdog** [comment]
Had some extra time, pulled from upstream and squashed the resulting bugs.
I also removed the custom BindingPatternExpression and replaced it with array and object literals, as requested.

I hope this can be merged soon!

**Comment by romdotdog** [comment]
Closing after the talk on [Discord](https://discord.com/channels/721472913886281818/874686570148536420/914344796431216670) about the `.identifierNode` 'hack' used throughout this PR, view by joining the [Discord Server](https://discord.gg/assemblyscript)
